interpreter/cling/lib/Interpreter/ExternalInterpreterSource.cpp

```diff
@@ -0,0 +1,302 @@
+      }
+      // Put the name of the Decl imported with the
+      // DeclarationName coming from the parent, in  my map.
+      m_ImportedDecls[childDeclName] = parentDeclName;
+    }
+  }
+
+  void ExternalInterpreterSource::ImportDeclContext(
+                                  DeclContext *declContextToImport,
+                                  ASTImporter &importer,
+                                  DeclarationName &childDeclName,
+                                  DeclarationName &parentDeclName,
+                                  const DeclContext *childCurrentDeclContext) {
+
+    if (DeclContext *importedDeclContext =
+                                importer.ImportContext(declContextToImport)) {
+
+      importedDeclContext->setHasExternalVisibleStorage(true);
+      if (NamedDecl *importedNamedDecl = 
+                            llvm::dyn_cast<NamedDecl>(importedDeclContext)) {
+        std::vector <NamedDecl*> declVector{importedNamedDecl};
+        llvm::ArrayRef <NamedDecl*> FoundDecls(declVector);
+        SetExternalVisibleDeclsForName(childCurrentDeclContext,
+                                       importedNamedDecl->getDeclName(),
+                                       FoundDecls);
+      }
+
+      // Put the name of the DeclContext imported with the
+      // DeclarationName coming from the parent, in  my map.
+      m_ImportedDecls[childDeclName] = parentDeclName;
+
+      // And also put the declaration context I found from the parent Interpreter
+      // in the map of the child Interpreter to have it for the future.
+      m_ImportedDeclContexts[importedDeclContext] = declContextToImport;
+    }
+  }
+
+  bool ExternalInterpreterSource::Import(DeclContext::lookup_result lookup_result,
+                                ASTContext &fromASTContext,
+                                ASTContext &toASTContext,
+                                const DeclContext *childCurrentDeclContext,
+                                DeclarationName &childDeclName,
+                                DeclarationName &parentDeclName) {
+
+    // Cling's ASTImporter
+    ClingASTImporter importer = CreateClingASTImporter(toASTContext,
+                                                       fromASTContext,
+                                                       m_ChildInterpreter,
+                                                       m_ParentInterpreter,
+                                                       *this);
+
+    for (DeclContext::lookup_iterator I = lookup_result.begin(),
+          E = lookup_result.end(); I != E; ++I) {
+      // Check if this Name we are looking for is
+      // a DeclContext (for example a Namespace, function etc.).
+      if (DeclContext *declContextToImport = llvm::dyn_cast<DeclContext>(*I)) {
+
+        ImportDeclContext(declContextToImport, importer, childDeclName,
+                          parentDeclName, childCurrentDeclContext);
+
+      } else if (Decl *declToImport = llvm::dyn_cast<Decl>(*I)) {
```

Comment:
Yes I realize that a lot of this is not your code. Thanks for your help in keeping other devs' code clean, too!